File: backend/services/ollamaService.js
/**
 * Ollama AI Service
 * Handles AI question generation using local Ollama instance
 */

const axios = require('axios');
const config = require('../config');
const logger = require('../utils/logger');

/**
 * Generate MCQ questions from syllabus content
 * @param {string} syllabusContent - Text content of syllabus
 * @param {number} questionCount - Number of questions to generate
 * @param {string} subject - Subject name for context
 * @returns {Promise<Array>} Array of generated questions
 */
const generateQuestions = async (syllabusContent, questionCount = 10, subject = '', difficulty = 'medium', questionType = 'MCQ') => {
  const prompt = buildQuestionPrompt(syllabusContent, questionCount, subject, difficulty, questionType);

  try {
    logger.info('Calling Ollama API for question generation', { questionCount, subject });

    const response = await axios.post(config.ollama.apiUrl, {
      model: config.ollama.model,
      prompt: prompt,
      format: "json",
      stream: false,
      options: {
        temperature: 0.7,
        top_p: 0.9,
        num_predict: 4096,
      },
    }, {
      timeout: 300000, // 5 minutes timeout for CPU AI generation
    });

    if (!response.data || !response.data.response) {
      throw new Error('Invalid response from Ollama');
    }

    const questions = parseQuestionsFromResponse(response.data.response, questionType);
    logger.info(`Generated ${questions.length} questions successfully`);

    return questions;
  } catch (error) {
    logger.error('Ollama API error:', {
      message: error.message,
      code: error.code,
    });

    if (error.code === 'ECONNREFUSED') {
      throw new Error('Ollama service is not running. Please start Ollama on localhost:11434');
    }

    if (error.code === 'ETIMEDOUT') {
      throw new Error('Question generation timed out. Please try with fewer questions.');
    }

    throw new Error(`Failed to generate questions: ${error.message}`);
  }
};

/**
 * Build the prompt for question generation
 * @param {string} content - Syllabus content
 * @param {number} count - Number of questions
 * @param {string} subject - Subject name
 * @returns {string} Formatted prompt
 */
const buildQuestionPrompt = (content, count, subject, difficulty, questionType) => {
  if (questionType === 'SUBJECTIVE') {
    return `You are an expert exam question creator. Generate ${count} SUBJECTIVE (descriptive/essay) questions based on the following syllabus/content for ${subject || 'the given subject'}.
    
IMPORTANT: The difficulty level of these questions MUST BE ${difficulty.toUpperCase()}. Adjust the complexity and depth of knowledge accordingly.

IMPORTANT: Return ONLY a valid JSON array. Do not include any text before or after the JSON.

Each question object must have this exact structure:
{
  "question_text": "The question text here?",
  "model_answer": "A detailed model answer or grading rubric that explains what a correct student response should contain."
}

Requirements:
- Generate exactly ${count} questions
- Questions should test deep understanding and require descriptive answers
- The difficulty level of all questions should be strictly ${difficulty.toUpperCase()}

SYLLABUS/CONTENT:
${content}

Return only the JSON array with ${count} questions:`;
  }

    return `You are an expert exam question creator. Generate exactly ${count} multiple choice questions (MCQs) based on the following syllabus/content for ${subject || 'the given subject'}.
  
IMPORTANT: The difficulty level of these questions MUST BE ${difficulty.toUpperCase()}.

CRITICAL INSTRUCTION: You MUST return ONLY a valid JSON array of ${count} objects. DO NOT output any introduction, explanation, or conversational text. ONLY JSON.

Each question object MUST have this EXACT structure with NO missing fields:
{
  "question_text": "The text of the question?",
  "option_a": "First option",
  "option_b": "Second option",
  "option_c": "Third option",
  "option_d": "Fourth option",
  "correct_option": "A"
}

Requirements:
- Generate exactly ${count} questions.
- Each question must have exactly 4 options (option_a, option_b, option_c, option_d).
- "correct_option" must be exactly "A", "B", "C", or "D".
- YOU MUST format it as a valid JSON array like [ { ... }, { ... } ].

SYLLABUS/CONTENT:
${content}

Return ONLY the JSON array:`;
};

/**
 * Parse questions from Ollama response
 * @param {string} responseText - Raw response text from Ollama
 * @returns {Array} Parsed questions array
 */
const parseQuestionsFromResponse = (responseText, questionType = 'MCQ') => {
  try {
    logger.info('Raw Ollama response:', { responseText });
    let questions = [];
    
    // Try to find JSON array in response
    const jsonMatch = responseText.match(/\[[\s\S]*\]/);
    if (jsonMatch) {
      questions = JSON.parse(jsonMatch[0]);
    } else {
      // If no array, try to find a single JSON object
      const singleMatch = responseText.match(/\{[\s\S]*\}/);
      if (singleMatch) {
        const parsed = JSON.parse(singleMatch[0]);
        // Sometimes it wraps the array in an object like {"questions": [...]}
        if (parsed.questions && Array.isArray(parsed.questions)) {
          questions = parsed.questions;
        } else {
          questions = [parsed];
        }
      } else {
        logger.warn('No JSON array or object found in Ollama response');
        return [];
      }
    }

    if (questionType === 'SUBJECTIVE') {
      const validQuestions = questions
        .filter(q => q.question_text && q.model_answer)
        .map(q => ({
          question: q.question_text.trim(),
          modelAnswer: q.model_answer.trim(),
          questionType: 'SUBJECTIVE',
          marks: 5,
        }));
      return validQuestions;
    }

    // Validate and clean questions
    const validQuestions = questions
      .filter(q => {
        const hasQuestion = q.question_text || q.questionText || q.question;
        const hasOptions = (q.option_a || q.optionA || q.a || (q.options && q.options[0])) && 
                           (q.option_b || q.optionB || q.b || (q.options && q.options[1])) && 
                           (q.option_c || q.optionC || q.c || (q.options && q.options[2])) && 
                           (q.option_d || q.optionD || q.d || (q.options && q.options[3]));
        return hasQuestion && hasOptions;
      })
      .map(q => {
        const correct = (q.correct_option || q.correctOption || q.correct_answer || q.correctAnswer || q.answer || 'A').toString().toUpperCase().trim();
        let correctIndex = 0;
        if (correct.includes('A') || correct === '0') correctIndex = 0;
        else if (correct.includes('B') || correct === '1') correctIndex = 1;
        else if (correct.includes('C') || correct === '2') correctIndex = 2;
        else if (correct.includes('D') || correct === '3') correctIndex = 3;

        return {
          question: (q.question_text || q.questionText || q.question).trim(),
          options: [
            (q.option_a || q.optionA || q.a || (q.options && q.options[0])).trim(),
            (q.option_b || q.optionB || q.b || (q.options && q.options[1])).trim(),
            (q.option_c || q.optionC || q.c || (q.options && q.options[2])).trim(),
            (q.option_d || q.optionD || q.d || (q.options && q.options[3])).trim()
          ],
          correctAnswer: correctIndex,
          questionType: 'MCQ',
          marks: 1,
        };
      });

    return validQuestions;
  } catch (error) {
    logger.error('Failed to parse Ollama response:', error.message);
    return [];
  }
};

/**
 * Check if Ollama service is available
 * @returns {Promise<boolean>} True if available
 */
const checkAvailability = async () => {
  try {
    const response = await axios.get(`${config.ollama.apiUrl.replace('/api/generate', '/api/tags')}`, {
      timeout: 5000,
    });
    return response.status === 200;
  } catch (error) {
    logger.warn('Ollama service not available:', error.message);
    return false;
  }
};

/**
 * Get list of available models
 * @returns {Promise<Array>} List of model names
 */
const getAvailableModels = async () => {
  try {
    const response = await axios.get(`${config.ollama.apiUrl.replace('/api/generate', '/api/tags')}`, {
      timeout: 5000,
    });
    return response.data.models?.map(m => m.name) || [];
  } catch (error) {
    logger.warn('Failed to get Ollama models:', error.message);
    return [];
  }
};

/**
 * Evaluate a subjective answer
 * @param {string} question - The question text
 * @param {string} modelAnswer - The expected model answer
 * @param {string} studentAnswer - The student's typed answer
 * @param {number} maxMarks - The maximum marks available for this question
 * @returns {Promise<{score: number, feedback: string}>}
 */
const evaluateSubjectiveAnswer = async (question, modelAnswer, studentAnswer, maxMarks = 5) => {
  const prompt = `You are an expert teacher grading an exam.
You must grade the student's answer to the following question.

QUESTION: ${question}
MODEL ANSWER (Expected Points): ${modelAnswer}
STUDENT ANSWER: ${studentAnswer || "(No answer provided)"}

You must assign a score from 0 to ${maxMarks} (integers or half marks). 
Also provide brief, constructive feedback on why this score was awarded.

IMPORTANT: Return ONLY a valid JSON object. Do not include any text before or after the JSON.
{
  "score": <number>,
  "feedback": "<string>"
}
`;

  try {
    const response = await axios.post(config.ollama.apiUrl, {
      model: config.ollama.model,
      prompt: prompt,
      stream: false,
      options: {
        temperature: 0.1, // Low temperature for more consistent grading
        top_p: 0.9,
      },
    }, {
      timeout: 60000, // 1 minute timeout per question
    });

    const jsonMatch = response.data.response.match(/\{[\s\S]*\}/);
    if (!jsonMatch) {
      throw new Error('No JSON object found in Ollama grading response');
    }

    const result = JSON.parse(jsonMatch[0]);
    return {
      score: Number(result.score) || 0,
      feedback: result.feedback || 'Graded by AI.'
    };
  } catch (error) {
    logger.error('Failed to grade subjective answer with Ollama:', error.message);
    return {
      score: 0,
      feedback: 'AI grading failed. Please review manually.'
    };
  }
};

module.exports = {
  generateQuestions,
  checkAvailability,
  getAvailableModels,
  evaluateSubjectiveAnswer,
};

File: backend-spring/src/main/java/com/examify/backend/service/OllamaService.java
package com.examify.backend.service;

import com.examify.backend.dto.TeacherDto.GeneratedQuestion;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Service
public class OllamaService {

    private final RestTemplate restTemplate = new RestTemplate();
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final String OLLAMA_URL = "http://localhost:11434/api/generate";
    private final String OLLAMA_MODEL = "llama3:latest"; // Or configured from properties

    public List<GeneratedQuestion> generateQuestions(String syllabusContent, int count, String subject, String defaultDifficulty, String questionType, String customPrompt) {
        List<GeneratedQuestion> allQuestions = new ArrayList<>();
        
        Map<String, Integer> difficultyCounts = extractDifficultyCounts(customPrompt);
        
        if (!difficultyCounts.isEmpty()) {
            int explicitCount = 0;
            for (Map.Entry<String, Integer> entry : difficultyCounts.entrySet()) {
                String diff = entry.getKey();
                int needed = entry.getValue();
                allQuestions.addAll(generateBatch(syllabusContent, needed, subject, diff, questionType, customPrompt));
                explicitCount += needed;
            }
            
            int remaining = count - explicitCount;
            if (remaining > 0) {
                 allQuestions.addAll(generateBatch(syllabusContent, remaining, subject, defaultDifficulty, questionType, customPrompt));
            }
        } else {
            allQuestions.addAll(generateBatch(syllabusContent, count, subject, defaultDifficulty, questionType, customPrompt));
        }
        
        if (allQuestions.size() > count) {
            return allQuestions.subList(0, count);
        }
        return allQuestions;
    }

    private List<GeneratedQuestion> generateBatch(String syllabusContent, int count, String subject, String difficulty, String questionType, String customPrompt) {
        if (count <= 0) return new ArrayList<>();
        
        List<GeneratedQuestion> allQuestions = new ArrayList<>();
        int attempts = 0;
        int maxAttempts = 3;

        while (allQuestions.size() < count && attempts < maxAttempts) {
            int needed = count - allQuestions.size();
            String prompt = buildPrompt(syllabusContent, needed, subject, difficulty, questionType, customPrompt);

            Map<String, Object> requestBody = new HashMap<>();
            requestBody.put("model", OLLAMA_MODEL);
            requestBody.put("prompt", prompt);
            requestBody.put("stream", false);
            requestBody.put("format", "json");

            Map<String, Object> options = new HashMap<>();
            options.put("temperature", 0.3);
            options.put("num_predict", -1);
            requestBody.put("options", options);

            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            HttpEntity<Map<String, Object>> entity = new HttpEntity<>(requestBody, headers);

            try {
                ResponseEntity<String> response = restTemplate.postForEntity(OLLAMA_URL, entity, String.class);
                JsonNode root = objectMapper.readTree(response.getBody());
                String responseText = root.path("response").asText();
                
                List<GeneratedQuestion> batch = parseQuestions(responseText, questionType);
                if (batch.isEmpty()) break;
                
                // Force difficulty to match the requested batch difficulty
                for (GeneratedQuestion q : batch) {
                    q.setDifficulty(difficulty);
                }
                
                allQuestions.addAll(batch);
            } catch (Exception e) {
                if (allQuestions.isEmpty()) {
                    throw new RuntimeException("Failed to generate questions via Ollama: " + e.getMessage(), e);
                } else {
                    break;
                }
            }
            attempts++;
        }
        return allQuestions;
    }

    private Map<String, Integer> extractDifficultyCounts(String customPrompt) {
        Map<String, Integer> counts = new HashMap<>();
        if (customPrompt == null || customPrompt.trim().isEmpty()) {
            return counts;
        }
        
        String lower = customPrompt.toLowerCase();
        int easy = extractCount(lower, "easy");
        int medium = extractCount(lower, "medium");
        int hard = extractCount(lower, "hard", "difficult");
        
        if (easy > 0) counts.put("Easy", easy);
        if (medium > 0) counts.put("Medium", medium);
        if (hard > 0) counts.put("Hard", hard);
        
        return counts;
    }
    
    private int extractCount(String text, String... keywords) {
        for (String kw : keywords) {
            Matcher m = Pattern.compile("(\\d+)\\s+" + kw).matcher(text);
            if (m.find()) {
                return Integer.parseInt(m.group(1));
            }
        }
        return 0;
    }

    private String buildPrompt(String syllabusContent, int count, String subject, String difficulty, String questionType, String customPrompt) {
        StringBuilder sb = new StringBuilder();
        
        sb.append("You are an expert AI teacher. You must generate EXACTLY ").append(count)
          .append(" ").append(difficulty).append(" ").append(questionType)
          .append(" questions.\n\n");
          
        if (customPrompt != null && !customPrompt.trim().isEmpty()) {
            sb.append("Additional Instructions from teacher:\n");
            sb.append(customPrompt).append("\n\n");
            sb.append("CRITICAL REQUIREMENT: Focus on generating EXACTLY ").append(count)
              .append(" ").append(difficulty).append(" questions regardless of the teacher's instructions above. DO NOT stop generating until you have reached exactly ").append(count).append(" questions!\n\n");
        }
        
        sb.append("Text:\n").append(syllabusContent).append("\n\n");

        sb.append("CRITICAL REQUIREMENT:\n");
        if ("SUBJECTIVE".equalsIgnoreCase(questionType)) {
            sb.append("You are an expert exam question creator.\n");
            sb.append("IMPORTANT: Return ONLY a valid JSON array. Do not include any text before or after the JSON.\n");
            sb.append("Each question object must have this exact structure:\n");
            sb.append("{\n");
            sb.append("  \"question_text\": \"The question text here?\",\n");
            sb.append("  \"difficulty\": \"Easy\",\n");
            sb.append("  \"model_answer\": \"A detailed model answer or grading rubric that explains what a correct student response should contain.\"\n");
            sb.append("}");
        } else {
            sb.append("You are an expert exam question creator.\n");
            sb.append("CRITICAL INSTRUCTION: You MUST return ONLY a valid JSON object containing a 'questions' array. DO NOT output any introduction, explanation, or conversational text. Start your response with { and end with }.\n");
            sb.append("WARNING: You MUST assign each option to the specific JSON keys \"option_a\", \"option_b\", \"option_c\", and \"option_d\". Do NOT output raw strings like \"A. Option\" inside the object!\n");
            sb.append("The JSON MUST exactly match this format:\n");
            sb.append("{\n");
            sb.append("  \"questions\": [\n");
            sb.append("    {\n");
            sb.append("      \"question_text\": \"What is the capital of France?\",\n");
            sb.append("      \"difficulty\": \"Easy\",\n");
            sb.append("      \"option_a\": \"Berlin\",\n");
            sb.append("      \"option_b\": \"Madrid\",\n");
            sb.append("      \"option_c\": \"Paris\",\n");
            sb.append("      \"option_d\": \"Rome\",\n");
            sb.append("      \"correct_option\": \"C\"\n");
            sb.append("    }\n");
            sb.append("  ]\n");
            sb.append("}");
        }
        
        return sb.toString();
    }

    private List<GeneratedQuestion> parseQuestions(String responseText, String questionType) throws JsonProcessingException {
        List<GeneratedQuestion> result = new ArrayList<>();
        System.out.println("Ollama Response: " + responseText);
        
        JsonNode root;
        try {
            root = objectMapper.readTree(responseText);
        } catch (Exception e) {
            // fallback: try to find JSON object in text
            Matcher objMatcher = Pattern.compile("\\{[\\s\\S]*\\}").matcher(responseText);
            if (objMatcher.find()) {
                root = objectMapper.readTree(objMatcher.group());
            } else {
                return result;
            }
        }

        JsonNode arrayNode = root;
        if (root.isObject() && root.has("questions")) {
            arrayNode = root.get("questions");
        } else if (root.isObject() && !root.has("questions")) {
            // single question
            arrayNode = objectMapper.createArrayNode().add(root);
        }

        if (!arrayNode.isArray()) {
            return result;
        }

        for (JsonNode qNode : arrayNode) {
            GeneratedQuestion q = new GeneratedQuestion();
            if ("SUBJECTIVE".equalsIgnoreCase(questionType)) {
                q.setQuestion(qNode.path("question_text").asText(qNode.path("question").asText()));
                q.setModelAnswer(qNode.path("model_answer").asText());
                q.setDifficulty(qNode.path("difficulty").asText("Medium"));
                q.setQuestionType("SUBJECTIVE");
                q.setMarks(5);
            } else {
                q.setQuestion(qNode.path("question_text").asText(qNode.path("question").asText()));
                q.setDifficulty(qNode.path("difficulty").asText("Medium"));
                List<String> options = new ArrayList<>();
                options.add(qNode.path("option_a").asText(qNode.path("a").asText()));
                options.add(qNode.path("option_b").asText(qNode.path("b").asText()));
                options.add(qNode.path("option_c").asText(qNode.path("c").asText()));
                options.add(qNode.path("option_d").asText(qNode.path("d").asText()));
                q.setOptions(options);
                
                String correctStr = qNode.path("correct_option").asText(qNode.path("correctAnswer").asText("A")).toUpperCase();
                int correctIndex = 0;
                if (correctStr.contains("B")) correctIndex = 1;
                else if (correctStr.contains("C")) correctIndex = 2;
                else if (correctStr.contains("D")) correctIndex = 3;
                
                q.setCorrectAnswer(correctIndex);
                q.setQuestionType("MCQ");
                q.setMarks(1);
            }
            result.add(q);
        }
        
        return result;
    }

    public Map<String, Object> evaluateSubjectiveAnswer(String question, String modelAnswer, String studentAnswer, int maxMarks) {
        String prompt = "You are an expert teacher grading an exam.\n" +
                "You must grade the student's answer to the following question.\n\n" +
                "QUESTION: " + question + "\n" +
                "MODEL ANSWER (Expected Points): " + modelAnswer + "\n" +
                "STUDENT ANSWER: " + (studentAnswer != null && !studentAnswer.isEmpty() ? studentAnswer : "(No answer provided)") + "\n\n" +
                "You must assign a score from 0 to " + maxMarks + " (integers or half marks). \n" +
                "Also provide brief, constructive feedback on why this score was awarded.\n\n" +
                "IMPORTANT: Return ONLY a valid JSON object. Do not include any text before or after the JSON.\n" +
                "{\n" +
                "  \"score\": <number>,\n" +
                "  \"feedback\": \"<string>\"\n" +
                "}\n";

        Map<String, Object> requestBody = new HashMap<>();
        requestBody.put("model", OLLAMA_MODEL);
        requestBody.put("prompt", prompt);
        requestBody.put("stream", false);
        requestBody.put("format", "json");

        Map<String, Object> options = new HashMap<>();
        options.put("temperature", 0.1); // Low temp for consistent grading
        requestBody.put("options", options);

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        HttpEntity<Map<String, Object>> entity = new HttpEntity<>(requestBody, headers);

        try {
            ResponseEntity<String> response = restTemplate.postForEntity(OLLAMA_URL, entity, String.class);
            JsonNode root = objectMapper.readTree(response.getBody());
            String responseText = root.path("response").asText();
            
            Matcher matcher = Pattern.compile("\\{[\\s\\S]*\\}").matcher(responseText);
            if (matcher.find()) {
                JsonNode evalNode = objectMapper.readTree(matcher.group());
                Map<String, Object> result = new HashMap<>();
                result.put("score", evalNode.path("score").asInt(0));
                result.put("feedback", evalNode.path("feedback").asText("Graded by AI."));
                return result;
            }
        } catch (Exception e) {
            System.err.println("AI grading failed: " + e.getMessage());
        }

        Map<String, Object> fallback = new HashMap<>();
        fallback.put("score", 0);
        fallback.put("feedback", "AI grading failed. Please review manually.");
        return fallback;
    }
}
